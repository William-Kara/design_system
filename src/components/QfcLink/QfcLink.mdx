import { ArgTypes, Meta } from '@storybook/blocks';
import QfcLink from '@/components/QfcLink/QfcLink.vue';

<Meta title="Elements/Link" />

# Definition

QfcLink component is typically used to create a clickable element that redirects
the user to a different page or URL when clicked. It can be used to navigate
between pages within a web application, or to open external links in a new window
or tab.

---

## Props

The following props are available for the QfcLink component:

<ArgTypes of={QfcLink} />

You may also use it without isExternalLink prop in that case it will use the default value which is false.

# Usage

To use QfcHeader, style needs to be imported separately.

```sass
@import '@website/design-system-vuejs/dist/styles/components/QfcLink.scss';
```
