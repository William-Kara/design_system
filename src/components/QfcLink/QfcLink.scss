.qfc-link {
  font-weight: $medium;
  font-size: 1.12rem; //m1
  color: $text;
  text-decoration: none;
  &:hover {
    color: $text;
    text-decoration: underline;
  }
  &--underline {
    text-decoration: underline;
  }
}
