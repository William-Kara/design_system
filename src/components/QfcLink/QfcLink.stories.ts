import Link from './QfcLink.vue';
import type { Meta, StoryObj } from '@storybook/vue3';

const meta: Meta<typeof Link> = {
  component: Link,
  title: 'Elements/Link',
  argTypes: {
    label: {
      control: {
        type: 'text',
      },
    },
    href: {
      control: {
        type: 'text',
      },
    },
    isExternalLink: {
      control: {
        type: 'boolean',
      },
    },
    underline: {
      control: {
        type: 'boolean',
      },
    },
  },
};
export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {
    label: 'Link',
    href: 'https://www.qongzi.com',
    isExternalLink: false,
    underline: false,
  },
};

export const Icon: Story = {
  args: {
    label: 'Link',
    href: 'https://www.qongzi.com',
    isExternalLink: false,
    iconSrc: '/image-q.svg',
    underline: false,
  },
};

export const Inline: Story = {
  args: {
    label: 'Link',
    href: 'https://www.qongzi.com',
    isExternalLink: false,
    underline: true,
  },
};
