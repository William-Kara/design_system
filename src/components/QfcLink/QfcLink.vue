<template>
  <span v-if="iconSrc">
    <a v-if="isExternalLink" :class="classes" :href="href" target="_blank">
      <img :src="iconSrc" width="24" height="24" alt="link icon" />
    </a>

    <router-link v-else :class="classes" :to="href" target="_self">
      <img :src="iconSrc" width="24" height="24" alt="link icon" />
    </router-link>
  </span>

  <span v-else>
    <a v-if="isExternalLink" :class="classes" :href="href" target="_blank">
      <slot>{{ label }}</slot>
    </a>

    <router-link v-else :class="classes" :to="href" target="_self">
      <slot>{{ label }}</slot>
    </router-link>
  </span>
</template>

<script lang="ts" setup>
type Props = {
  label?: string;
  href: string;
  isExternalLink?: boolean;
  iconSrc: string;
  underline?: boolean;
};

const props = withDefaults(defineProps<Props>(), {
  isExternalLink: false,
});

const classes = {
  'qfc-link': true,
  'qfc-link--underline': props.underline,
};
</script>

<style lang="scss" scoped src="./QfcLink.scss" />
