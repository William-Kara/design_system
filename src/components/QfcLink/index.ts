import QfcLink from './QfcLink.vue';

export { QfcLink };
export default QfcLink;
