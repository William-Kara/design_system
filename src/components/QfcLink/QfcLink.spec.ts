import { it, describe, beforeEach, expect } from 'vitest';

import { shallowMount } from '@vue/test-utils';
import QfcLink from './QfcLink.vue';

let wrapper: any = null;

const propsValues = {
  label: 'Link',
  href: 'https://www.qongzi.com',
  isExternalLink: true,
  underline: false,
};

beforeEach(() => {
  wrapper = shallowMount(QfcLink, {
    propsData: propsValues,
  });
});

describe('QfcLink Component', () => {
  it('should mount properly', () => {
    expect(wrapper).toBeTruthy();
  });

  it('should have correct props', () => {
    expect(wrapper.props()).toEqual(propsValues);
  });

  it('should render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render router-link when isExternalLink is false', () => {
    const props = {
      label: 'Link',
      href: 'https://www.qongzi.com',
      isExternalLink: false,
    };

    const wrapper = shallowMount(QfcLink, {
      propsData: props,
    });

    expect(wrapper.find('router-link').exists()).toBe(true);
  });

  it('should render a tag when isExternalLink is true', () => {
    expect(wrapper.props().isExternalLink).toBe(true);
    expect(wrapper.find('a').exists()).toBe(true);
  });

  it('sets target to _blank when isExternalLink is true', () => {
    expect(wrapper.props().isExternalLink).toBe(true);
    const link = wrapper.find('a');
    expect(link.attributes('target')).toBe('_blank');
  });

  it('sets target to _self when isExternalLink is false', () => {
    const props = {
      label: 'Link',
      href: 'https://www.qongzi.com',
      isExternalLink: false,
    };

    const wrapper = shallowMount(QfcLink, {
      propsData: props,
    });
    const link = wrapper.find('router-link');
    expect(link.attributes().target).toBe('_self');
  });
});
