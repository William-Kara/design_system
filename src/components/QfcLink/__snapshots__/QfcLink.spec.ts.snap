// Vitest Snapshot v1

exports[`QfcLink Component > should render correctly 1`] = `"<span data-v-a949817b=\\"\\"><a data-v-a949817b=\\"\\" class=\\"qfc-link\\" href=\\"https://www.qongzi.com\\" target=\\"_blank\\">Link</a></span>"`;
